### PR TITLE
[d3d9] Use SSBOs for SWVP if we extend out of uniformRange

### PR DESCRIPTION
**Pull request by Joshua-Ashton**

### CODE REVIEW

**Comment by Joshua-Ashton** [comment]
No, only when SWVPing

**Comment by doitsujin** [comment]
Where exactly the check for that?

In that case I'd recommend **always** using an SSBO for SWVP for simplicity. On AMD there's no difference at all between UBOs and SSBOs, and I don't think Intel cares very much either.

File: src/d3d9/d3d9_device.cpp [hunk 1]
```diff
@@ -4432,69 +4432,84 @@ namespace dxvk {
   }
 
 
-  void D3D9DeviceEx::CreateConstantBuffers() {
-    DxvkBufferCreateInfo info;
-    info.usage  = VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT;
-    info.access = VK_ACCESS_UNIFORM_READ_BIT;
+  Rc<DxvkBuffer> D3D9DeviceEx::CreateConstantBuffer(
+          bool                SSBO,
```
[the remark this answers is not in the document]
Comment (Joshua-Ashton) [reply src/d3d9/d3d9_device.cpp]: I would prefer it this way cause then I don't need to do 2 ternaries for the usage + access when creating the vertex constant buffer.

Either that or some other determiner for it being an SSBO.

**Comment by Joshua-Ashton** [comment]
> Where exactly the check for that?

dxso_options.cpp

``vertexConstantBufferAsSSBO = pDevice->GetVertexConstantLayout().totalSize() > devInfo.core.properties.limits.maxUniformBufferRange;``

> In that case I'd recommend always using an SSBO for SWVP [...]

If we do it this way then we can still use UBOs if the max constant limits are set to more sane values in dxvk.conf (like we do for Sims 2 perf)

**Comment by doitsujin** [comment]
```
vertexConstantBufferAsSSBO = pDevice->GetVertexConstantLayout().totalSize() > devInfo.core.properties.limits.maxUniformBufferRange;
```
I don't see how this **only** affects SWVP draw calls. The values returned by `GetVertexConstantLayout` are set during device initialization based on device create flags.

**Comment by Joshua-Ashton** [comment]
Those device create flags determine whether we SWVP or not

**Comment by doitsujin** [comment]
Wait, is that device global and not just for `ProcessVertices` shenanigans?

**Comment by Joshua-Ashton** [comment]
Yep, ProcessVertices is separate entirely (that just runs any given shader in "software" and outputs to a constant buffer)

SWVP in D3D9 is, hey here's 8192 vec4 constants, 2048 ivec4 constants and 2048 bool constants, draw with them 🐸

**Pull request merged**